File: lua/minigrep.lua
-- Minigrep in Lua

local function get_matches(query, lines)
	local matches = {}
	local i = 0
	for line in lines do
		if string.find(line, query, 1, true) ~= nil then
			table.insert(matches, { i, line })
		end
		i = i + 1
	end
	return matches
end

local function print_match(match)
	local i, line = unpack(match)
	print(i .. ": " .. line)
end

local function print_matches(matches)
	for _, match_elem in ipairs(matches) do
		print_match(match_elem)
	end
end

local function grep(query, filename)
	local file = io.open(filename, "r")
	if not file then
		print("File not found")
	else
		local lines = file:lines()
		local matches = get_matches(query, lines)
		file:close()
		if #matches ~= 0 then
			print_matches(matches)
		else
			print("No matches found")
		end
	end
end

local function minigrep(query, filename)
	if query ~= nil or filename ~= nil then
		return grep(query, filename)
	else
		return print("Usage: minigrep <query> <filename>")
	end
end

local args = { ... }
minigrep(args[1], args[2])

File: nim/minigrep.nim
#[
  Minigrep in Nim
]#

import std/os, std/strutils, std/strformat

type Line = tuple[index: int, line: string]

proc get_matches(query: string, lines: seq[string]): seq[Line] =
  var
    matches: seq[Line]
    i = 0
  for line in lines:
    if line.find(query) != -1:
      matches.add((index: i, line: line))
    i += 1
  return matches

proc print_match(match: Line) =
  let (i, line) = match
  echo fmt"{i}: {line}"

proc print_matches(matches: seq[Line]) =
  for match in matches:
    print_match(match)

proc grep(query: string, filename: string) =
  let file = open(filename, fmRead)
  if file != nil:
    let lines = file.read_all().split_lines()
    let matches = get_matches(query, lines)
    file.close()
    if matches.len() != 0:
      print_matches(matches)
    else:
      echo "No matches found"
  else:
    echo "File not found"

proc minigrep(query: string, filename: string) =
  if filename == "" and query == "":
    echo "Usage: minigrep <query> <filename>"
  else:
    grep(query, filename)

minigrep(param_str(1), param_str(2))
